### PR TITLE
[OCPCLOUD-923] add alerts for memory and cpu core limits

### PR DESCRIPTION
**Pull request by elmiko**
This change adds the alerts for when the cluster autoscaler is unable to
scale out due to reaching cpu or memory limits. It also updates the
alert documents.

ref: https://issues.redhat.com/browse/OCPCLOUD-923

### CODE REVIEW

**Comment by elmiko** [comment]
this will need to wait until we do the 1.22 rebase of the cluster autoscaler to pick up the new metrics sources.
/hold

File: docs/user/alerts.md
Comment (RiRa12621) [inline docs/user/alerts.md:85]: Should a potentially non-harmful condition maybe drop down to info instead of warning?

Comment (elmiko) [reply docs/user/alerts.md:85]: i was curious about the same thing myself. if the autoscaler has reached its maximum (cpu or memory), it's definitely not an emergency in the cluster, but it might be something that a cluster admin would want to know since it could cause unexpected delays for users (pending pods and whatnot). does that sound more like an info level?

Comment (RiRa12621) [reply docs/user/alerts.md:85]: I'll take a peak what we have around pending pods already. If that is covered with an alert itself then an info level here should be good enough as supporting information. It still pops up in the console

Comment (elmiko) [reply docs/user/alerts.md:85]: awesome, thanks!

**Comment by dofinn** [comment]
awesome job on this @elmiko +1

File: docs/user/alerts.md
Comment (RiRa12621) [reply docs/user/alerts.md:85]: yeah i think 
```
   - alert: KubePodNotReady
      annotations:
        description: Pod {{ $labels.namespace }}/{{ $labels.pod }} has been in a non-ready state for longer than 15 minutes.
        summary: Pod has been in a non-ready state for more than 15 minutes.
      expr: |
        sum by (namespace, pod) (
          max by(namespace, pod) (
            kube_pod_status_phase{namespace=~"(openshift-.*|kube-.*|default|logging)",job="kube-state-metrics", phase=~"Pending|Unknown"}
          ) * on(namespace, pod) group_left(owner_kind) topk by(namespace, pod) (
            1, max by(namespace, pod, owner_kind) (kube_pod_owner{owner_kind!="Job"})
          )
        ) > 0
      for: 15m
      labels:
        severity: warning
        ```
        should be covering that case good enough

Comment (elmiko) [reply docs/user/alerts.md:85]: thanks for the confirmation @RiRa12621 , i will drop these to info level

**Comment by elmiko** [comment]
updated to drop severity from warning to info

**Comment by RiRa12621** [comment]
Lgtm from our end

**Review by michaelgugino — APPROVED** [review]
/lgtm

Info is definitely the right severity level for this one.

File: docs/user/alerts.md
Comment (JoelSpeed) [inline docs/user/alerts.md:67]: Is this default correct? Seems absurdly high? Do we have a source for that value? If so, is it worth adding a link inline?

File: docs/user/alerts.md
Comment (JoelSpeed) [inline docs/user/alerts.md:93]: Same as above, where did this figure come from?

**Review by JoelSpeed — COMMENTED** [review]
This seems good to me, just curious about the default values mentioned, PTAL

File: docs/user/alerts.md
Comment (elmiko) [reply docs/user/alerts.md:67]: this is just the default if you set nothing on the cluster autoscaler, see [this faq entry](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-are-the-parameters-to-ca), also we do allow [not setting those limits in a ClusterAutoscaler](https://github.com/openshift/cluster-autoscaler-operator/blob/master/pkg/apis/autoscaling/v1/clusterautoscaler_types.go#L77-L94)

File: docs/user/alerts.md
Comment (elmiko) [reply docs/user/alerts.md:93]: same as above

**Comment by elmiko** [comment]
reminder: update https://github.com/openshift/enhancements/pull/538 when this merges

File: docs/user/alerts.md
Comment (wanghaoran1988) [reply docs/user/alerts.md:67]: Could you add a description how you can get the current maxmum settings? like "You can run this to get what's the maximum cores setup in this cluster".

Comment (JoelSpeed) [reply docs/user/alerts.md:67]: @wanghaoran1988 Does this need to happen before we merge or could that follow in a later release? I'm happy with this personally and would prefer to merge it today so its in before feature freeze, but Mike is out on PTO today so wouldn't have time to update this before FF

Comment (wanghaoran1988) [reply docs/user/alerts.md:67]: I'm ok with a follow release.

**Comment by elmiko** [comment]
this PR technically needs the metric to be present, we will pick it up during the 1.22 rebase. i am removing the hold as we don't believe this will break anything by merging early.
/hold cancel

**Comment by elmiko** [comment]
seeing this error several times in the logs
```
 03   11966 framework.go:128] error querying api for ValidatingWebhookConfiguration: no matches for kind "ValidatingWebhookConfiguration" in version "admissionregistration.k8s.io/v1beta1", retrying...
E0725 14:49:30.977789   11966 framework.go:128] error querying api for ValidatingWebhookConfiguration: no matches for kind "ValidatingWebhookConfiguration" in version "admissionregistration.k8s.io/v1beta1", retrying... 
```

doesn't seem related to this PR, but makes me wonder if that is one of the types that got upgraded to `v1` version?

**Comment by elmiko** [comment]
seems like yes, https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22

**Comment by JoelSpeed** [comment]
The webhook test is a known perma failure at the moment, we have a PR to our tests updating the webhook versions but other tests are flaking causing it not have merged yet

Having reviewed the failures, I don't think any of them are related to this PR, this should be safe to merge 

/override ci/prow/e2e-aws-operator

**Comment by elmiko** [comment]
thanks Joel, i was just looking at that webhook version failure lol

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-aws-operator

Re overide based on the above comment

**Pull request merged**
